### PR TITLE
Remove null values

### PR DESCRIPTION
**Pull request by abdulrahuman5196**
### Details
When Onyx.set() is called with null value we are storing it in OnyxCache and AsyncStorage without removing the data from the both. This is leading to iOS crashes for AsyncStorage.multiMerge() operation. More information can be found at the root cause analysis here - https://github.com/Expensify/App/issues/13907#issuecomment-1368779185
Reviewer - @jasperhuangg 

### Related Issues
https://github.com/Expensify/App/issues/13907

### Automated Tests
I have added added test naming 'should remove key value from OnyxCache/Storage when set is called with null value' which verifies if the data is removed from OnyxCache and Storage.

### Linked PRs
None. Currently Expensify is using '1.0.29' which is lower than current onyx version '1.0.32'. We have to confirm if we can update Expensify package.json with change or should we wait. Anyways it shouldn't block this PR.

### CODE REVIEW

**Comment by abdulrahuman5196** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by abdulrahuman5196** [comment]
recheck

**Comment by abdulrahuman5196** [comment]
recheck

File: lib/Onyx.js [hunk 1]
```diff
@@ -847,6 +846,11 @@ function evictStorageAndRetry(error, onyxMethod, ...args) {
  * @returns {Promise}
  */
 function set(key, value) {
+    if (_.isNull(value)) {
+        Logger.logInfo(`Removing the key: ${key} as it called set() with null value`);
```
Comment (tgolen) [inline lib/Onyx.js:850]: I think we should remove this log line, and also the log a couple of lines below. I don't think they are useful at all.

File: tests/unit/onyxTest.js [hunk 2]
```diff
@@ -24,11 +24,37 @@ Onyx.init({
 describe('Onyx', () => {
     let connectionID;
 
+    /** @type OnyxCache */
+    let cache;
+
+    beforeEach(() => {
+        cache = require('../../lib/OnyxCache').default;
```
Comment (tgolen) [inline tests/unit/onyxTest.js:31]: Why is this required for each test and not imported at the top of the file?

**Review by tgolen — CHANGES_REQUESTED** [review]

File: lib/Onyx.js (hunk 1 above)
Comment (abdulrahuman5196) [reply lib/Onyx.js:850]: >also the log a couple of lines below
To confirm. Do you mean this line

'Logger.logInfo(`set() called for key: ${key}${_.isObject(value) ? ` properties: ${_.keys(value).join(',')}` : ''}`);'  850/854? 

I added the above log line w.r.t the current log line present 850/854. I can remove both if they are not required.

File: tests/unit/onyxTest.js (hunk 2 above)
Comment (abdulrahuman5196) [reply tests/unit/onyxTest.js:31]: The OnyxCache resolves to undefined if we import it directly from the top of the file. I took this reference of using require() from existing unit tests found here and used same test logic.

https://github.com/Expensify/react-native-onyx/blob/7d6cdc339acc17aa90f50f61e3e825364163146b/tests/unit/onyxCacheTest.js#L17
https://github.com/Expensify/react-native-onyx/blob/af02f32b74ae7c870b75e029d3a16fe4079442cd/tests/unit/onyxClearWebStorageTest.js#L32
https://github.com/Expensify/react-native-onyx/blob/5686d24c198204d023a78c0780bb37850335cce6/tests/unit/onyxClearNativeStorageTest.js#L25

File: lib/Onyx.js (hunk 1 above)
Comment (abdulrahuman5196) [reply lib/Onyx.js]: Removed the log lines and added a new commit

Comment (tgolen) [reply lib/Onyx.js]: Yes, let's please remove both. I've been meaning to remove that existing log line for a while.

File: tests/unit/onyxTest.js (hunk 2 above)
Comment (tgolen) [reply tests/unit/onyxTest.js:31]: Ah, that's interesting. If other tests are doing it, I guess that's OK for now. Thanks!

File: lib/Onyx.js (hunk 1 above)
Comment (abdulrahuman5196) [reply lib/Onyx.js]: Done. Removed the log lines with the latest commit.

File: tests/unit/onyxTest.js [hunk 3]
```diff
@@ -24,11 +24,37 @@ Onyx.init({
...
+    });
+
     afterEach(() => {
         Onyx.disconnect(connectionID);
         return Onyx.clear();
     });
 
+    it('should remove key value from OnyxCache/Storage when set is called with null value',
+        () => Onyx.set(ONYX_KEYS.OTHER_TEST, 42)
```
Comment (tgolen) [inline tests/unit/onyxTest.js:40]:
```suggestion
    it('should remove key value from OnyxCache/Storage when set is called with null value', () => Onyx.set(ONYX_KEYS.OTHER_TEST, 42)
```
Just to match the style elsewhere.

**Review by tgolen — CHANGES_REQUESTED** [review]
Just this one minor change, please!

**Review by amyevans — APPROVED** [review]
@tgolen Do we have a process for back-porting patches to older versions (e.g. 1.0.29)? Perhaps in this case it doesn't really matter and we can just hold until we're ready to use the newest release in App. But also just curious in general 😄

**Comment by tgolen** [comment]
We don't, no. I think we should just HOLD this until the SQLite change is live in App, then it shouldn't really be an issue. It probably won't be an issue anyway (is my guess), but I am just being a little paranoid since there are a few changes in flux right now. I think really all that means is that maybe this won't get merged until Monday. I'm expecting that the SQLite change will pass testing and be merged today.

File: tests/unit/onyxTest.js (hunk 3 above)
Comment (abdulrahuman5196) [reply tests/unit/onyxTest.js]: Doen with latest commit

File: tests/unit/onyxTest.js [hunk 4]
```diff
@@ -24,11 +24,36 @@ Onyx.init({
...
+        .then(() => {
+            // Checks if cache value is removed.
+            expect(cache.getAllKeys().length).toBe(0);
+
+            // When cache keys length is 0, we fetch the keys from storage.
+            return Onyx.getAllKeys();
+        })
+        .then((keys) => {
+            expect(keys.includes(ONYX_KEYS.OTHER_TEST)).toBe(false);
+            return Onyx.set(ONYX_KEYS.OTHER_TEST, 42);
```
Comment (tgolen) [inline tests/unit/onyxTest.js:54]: Is this necessary? I don't think this value is ever verified anywhere, and the test is over at this point and then Onyx is cleared in `afterEach()`

**Review by tgolen — APPROVED** [review]

File: tests/unit/onyxTest.js (hunk 4 above)
Comment (abdulrahuman5196) [reply tests/unit/onyxTest.js]: I don't think its absolutely necessary. I just added it to reset the OTHER_TEST state which i removed. 
Anyways added a commit to remove it since removing it doesn't have any effect.

**Comment by abdulrahuman5196** [comment]
Thank you. @tgolen  and @amyevans  for reviewing and approving the PR quickly. I think this PR is good to be merged from your end(I wont have permission) after SQLite changes comes in. Kindly let me know if I have to do anything from my end.

**Review by jasperhuangg — APPROVED** [review]
looks good! nice job adding tests too :)

**Review by amyevans — APPROVED** [review]

**Comment by tgolen** [comment]
This is off HOLD and I've merged it. @abdulrahuman5196 you will need to do a PR for App to bump the version of Onyx and test it out there.

**Comment by abdulrahuman5196** [comment]
@tgolen  Thank you. Created PR to bump version - https://github.com/Expensify/App/pull/14178
